### DISCUSSION TITLE
FXGL for admin/business UI

### DISCUSSION THREAD

**Discussion by Higgins328** [AlmasB/FXGL#1155] [Q&A]
Hi,

having done a few UI:s in Java FX over the past years I ran into FXGL and see that it is able to perform a lot nicer UI experiences compared to plain vanilla FX. What I started out with now is a simple admin UI where the user first sees a splash screen, then a log in screen with user name and pw, and at last the actual admin form where some choices needs to be made by the user. So really two screens in total. I haven't seen any example out there for a business or admin form app using FXGL. 

What is the best approach to use? From what I found out after a few hours of trial and errors the addUINode seems to be the way to go. When I first used spawn I find it difficult to add event handlers to the spawned objects. 

This is my code so far, showing the essential part initGame. The drawback with using addUINode for all input controls is that it seems to be necessary to declare them all in initGame(). So when adding more screens to the app the initGame() procedure will be rather hefty

@Override
    protected void initGame() {

        TextField txtUserName = fxHelper.getStandardTextField("");
        txtUserName.setOnKeyTyped(event -> _txtUserPwValue = txtUserName.getText());

        TextField txtUserPw = fxHelper.getStandardTextField("");
        txtUserPw.setOnKeyTyped(event -> _txtUserNameValue = txtUserPw.getText());

        Button btnLogin = fxHelper.getStandardButton("Login");
        btnLogin.setOnAction(event -> btnLogin_onClick());

        AtomicInteger spawnCounter = new AtomicInteger();
        AtomicBoolean controlsSpawned = new AtomicBoolean(false);

        getGameScene().setBackgroundColor(Color.BLACK);
        getGameWorld().addEntityFactory(new Factory());

        runOnce(() -> {
            spawn("splashtext", 300, 500);
        }, Duration.seconds(0));

        run(() -> {

            spawnCounter.getAndIncrement();
            if (!controlsSpawned.get() && spawnCounter.compareAndSet(6, 14) ) {
                addUINode(txtUserName, 900, 700);
                addUINode(txtUserPw, 900, 770);
                addUINode(btnLogin, 900, 840);
                controlsSpawned.set(true);
            }

        }, Duration.seconds(0.50));
    }

Regards,
Henrik

**Comment by AlmasB** [comment]
Hi, FXGL behaves in a very similar way to a normal JavaFX app. If the only thing you are planning to use from FXGL is the UI, there is really no need to use entity factories and the like -- this can be done through normal Java / JavaFX code.

You can define each screen as a subscene. Check out the [wiki](https://github.com/AlmasB/FXGL/wiki/Scenes-SubScenes-Screens-%28FXGL-11%29).

Having a normal business application with FXGL is very feasible and I'm happy to add a new sample demonstrating it. If you could provide a full use case of the application, I'll open an issue to create a simple clone of that.

**Reply by Higgins328** [reply]
What I hope to achieve by using the FXGL UI is foremost the ability to move objects smoothly on the screen. So I guess the UI part from FXGL is what I want to use. The app will run on a raspberry pi with Ubuntu. There will be a MySQL or MariaDB backend to store user input, and a gRPC integration with a remote server for communication with other parts like smartphone app.

For this app, the use case will be user first sees a splash screen, then a login form with user name, password and a button to commit login. Credentials are validated towards database. If successful validation, next screen is loaded where user can edit various settings with some text fields, dropdowns, check boxes, radio buttons and so on.

I you could create a sample it would be most welcome!

**Reply by AlmasB** [reply]
Hi, not sure whether it is still useful, but I've finally put together a quick sample for a [business app](https://github.com/AlmasB/FXGL/blob/12e2cb4664a9ae93ff58eb9e3837ac5eeea381c7/fxgl-samples/src/main/java/sandbox/BusinessApp.java)

**Comment by Higgins328** [comment]
Great, thank you! Will look into it and see how to move on
